File: tests/algorithms/x/PHP/conversions/weight_conversion.php
<?php
error_reporting(E_ALL & ~E_DEPRECATED);
ini_set('memory_limit', '-1');
$now_seed = 0;
$now_seeded = false;
$s = getenv('MOCHI_NOW_SEED');
if ($s !== false && $s !== '') {
    $now_seed = intval($s);
    $now_seeded = true;
}
function _now() {
    global $now_seed, $now_seeded;
    if ($now_seeded) {
        $now_seed = ($now_seed * 1664525 + 1013904223) % 2147483647;
        return $now_seed;
    }
    return hrtime(true);
}
$__start_mem = memory_get_usage();
$__start = _now();
  function pow10($exp) {
  global $KILOGRAM_CHART, $WEIGHT_TYPE_CHART;
  $result = 1.0;
  if ($exp >= 0) {
  $i = 0;
  while ($i < $exp) {
  $result = $result * 10.0;
  $i = $i + 1;
};
} else {
  $i = 0;
  while ($i < (0 - $exp)) {
  $result = $result / 10.0;
  $i = $i + 1;
};
}
  return $result;
};
  $KILOGRAM_CHART = ['atomic-mass-unit' => 602213665200000036747951526.19481086730957, 'carrat' => 5000.0, 'gram' => 1000.0, 'kilogram' => 1.0, 'long-ton' => 0.0009842073, 'metric-ton' => 0.001, 'milligram' => 1000000.0, 'ounce' => 35.273990723, 'pound' => 2.2046244202, 'short-ton' => 0.0011023122, 'stone' => 0.1574731728];
  $WEIGHT_TYPE_CHART = ['atomic-mass-unit' => 1.660540199 * pow10(-27), 'carrat' => 0.0002, 'gram' => 0.001, 'kilogram' => 1.0, 'long-ton' => 1016.04608, 'metric-ton' => 1000.0, 'milligram' => 0.000001, 'ounce' => 0.0283495, 'pound' => 0.453592, 'short-ton' => 907.184, 'stone' => 6.35029];
  function weight_conversion($from_type, $to_type, $value) {
  global $KILOGRAM_CHART, $WEIGHT_TYPE_CHART;
  $has_to = array_key_exists($to_type, $KILOGRAM_CHART);
  $has_from = array_key_exists($from_type, $WEIGHT_TYPE_CHART);
  if ($has_to && $has_from) {
  return $value * $KILOGRAM_CHART[$to_type] * $WEIGHT_TYPE_CHART[$from_type];
}
  echo rtrim('Invalid \'from_type\' or \'to_type\''), PHP_EOL;
  return 0.0;
};
  echo rtrim(json_encode(weight_conversion('kilogram', 'gram', 1.0), 1344)), PHP_EOL;
  echo rtrim(json_encode(weight_conversion('gram', 'pound', 3.0), 1344)), PHP_EOL;
  echo rtrim(json_encode(weight_conversion('ounce', 'kilogram', 3.0), 1344)), PHP_EOL;
$__end = _now();
$__end_mem = memory_get_peak_usage(true);
$__duration = max(1, intdiv($__end - $__start, 1000));
$__mem_diff = max(0, $__end_mem - $__start_mem);
$__bench = ["duration_us" => $__duration, "memory_bytes" => $__mem_diff, "name" => "main"];
$__j = json_encode($__bench, 128);
$__j = str_replace("    ", "  ", $__j);
echo $__j, PHP_EOL;

File: tests/algorithms/x/Pascal/conversions/weight_conversion.pas
{$mode objfpc}
program Main;
uses SysUtils, fgl;
var _nowSeed: int64 = 0;
var _nowSeeded: boolean = false;
procedure init_now();
var s: string; v: int64;
begin
  s := GetEnvironmentVariable('MOCHI_NOW_SEED');
  if s <> '' then begin
    Val(s, v);
    _nowSeed := v;
    _nowSeeded := true;
  end;
end;
function _now(): integer;
begin
  if _nowSeeded then begin
    _nowSeed := (_nowSeed * 1664525 + 1013904223) mod 2147483647;
    _now := _nowSeed;
  end else begin
    _now := Integer(GetTickCount64()*1000);
  end;
end;
function _bench_now(): int64;
begin
  _bench_now := GetTickCount64()*1000;
end;
function _mem(): int64;
var h: TFPCHeapStatus;
begin
  h := GetFPCHeapStatus;
  _mem := h.CurrHeapUsed;
end;
procedure panic(msg: string);
begin
  writeln(msg);
  halt(1);
end;
var
  bench_start_0: integer;
  bench_dur_0: integer;
  bench_mem_0: int64;
  bench_memdiff_0: int64;
  KILOGRAM_CHART: specialize TFPGMap<string, real>;
  WEIGHT_TYPE_CHART: specialize TFPGMap<string, real>;
  exp: integer;
  value: real;
  to_type: string;
  from_type: string;
function Map1(): specialize TFPGMap<string, real>; forward;
function pow10(exp: integer): real; forward;
function weight_conversion(from_type: string; to_type: string; value: real): real; forward;
function Map1(): specialize TFPGMap<string, real>;
begin
  Result := specialize TFPGMap<string, real>.Create();
  Result.AddOrSetData('kilogram', Variant(1));
  Result.AddOrSetData('gram', Variant(1000));
  Result.AddOrSetData('milligram', Variant(1e+06));
  Result.AddOrSetData('metric-ton', Variant(0.001));
  Result.AddOrSetData('long-ton', Variant(0.0009842073));
  Result.AddOrSetData('short-ton', Variant(0.0011023122));
  Result.AddOrSetData('pound', Variant(2.2046244202));
  Result.AddOrSetData('stone', Variant(0.1574731728));
  Result.AddOrSetData('ounce', Variant(35.273990723));
  Result.AddOrSetData('carrat', Variant(5000));
  Result.AddOrSetData('atomic-mass-unit', Variant(6.022136652 * pow10(26)));
end;
function pow10(exp: integer): real;
var
  pow10_result_: real;
  pow10_i: integer;
begin
  pow10_result_ := 1;
  if exp >= 0 then begin
  pow10_i := 0;
  while pow10_i < exp do begin
  pow10_result_ := pow10_result_ * 10;
  pow10_i := pow10_i + 1;
end;
end else begin
  pow10_i := 0;
  while pow10_i < (0 - exp) do begin
  pow10_result_ := pow10_result_ / 10;
  pow10_i := pow10_i + 1;
end;
end;
  exit(pow10_result_);
end;
function weight_conversion(from_type: string; to_type: string; value: real): real;
var
  weight_conversion_has_to: boolean;
  weight_conversion_has_from: boolean;
begin
  weight_conversion_has_to := KILOGRAM_CHART.IndexOf(to_type) <> -1;
  weight_conversion_has_from := WEIGHT_TYPE_CHART.IndexOf(from_type) <> -1;
  if weight_conversion_has_to and weight_conversion_has_from then begin
  exit((value * KILOGRAM_CHART[to_type]) * WEIGHT_TYPE_CHART[from_type]);
end;
  writeln('Invalid ''from_type'' or ''to_type''');
  exit(0);
end;
begin
  init_now();
  bench_mem_0 := _mem();
  bench_start_0 := _bench_now();
  KILOGRAM_CHART := Map1();
  WEIGHT_TYPE_CHART := Map1();
  writeln(weight_conversion('kilogram', 'gram', 1));
  writeln(weight_conversion('gram', 'pound', 3));
  writeln(weight_conversion('ounce', 'kilogram', 3));
  bench_memdiff_0 := _mem() - bench_mem_0;
  bench_dur_0 := (_bench_now() - bench_start_0) div 1000;
  writeln('{');
  writeln(('  "duration_us": ' + IntToStr(bench_dur_0)) + ',');
  writeln(('  "memory_bytes": ' + IntToStr(bench_memdiff_0)) + ',');
  writeln(('  "name": "' + 'main') + '"');
  writeln('}');
end.
